File: storage.js
const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const path = require('path');

/**
 * Initializes and returns the Supabase client
 * using SUPABASE_URL and SUPABASE_KEY from environment variables.
 *
 * @returns {import('@supabase/supabase-js').SupabaseClient}
 */
function getSupabaseClient() {
  const supabaseUrl = process.env.SUPABASE_URL;
  const supabaseKey = process.env.SUPABASE_KEY;

  if (!supabaseUrl || !supabaseKey) {
    throw new Error('SUPABASE_URL and SUPABASE_KEY environment variables are required.');
  }

  return createClient(supabaseUrl, supabaseKey);
}

/**
 * Uploads a file to the Supabase 'uploads' storage bucket.
 *
 * @param {string|Buffer|Uint8Array|ArrayBuffer|ReadableStream} file - Local file path string, Buffer, or stream.
 * @param {string} [destinationPath] - Remote file path/name in the 'uploads' bucket.
 * @param {Object} [options] - Optional upload configuration.
 * @param {string} [options.bucket='uploads'] - Bucket name (defaults to 'uploads').
 * @param {string} [options.contentType] - MIME content type (e.g. 'application/pdf').
 * @param {boolean} [options.upsert=false] - Whether to overwrite an existing file with the same name.
 * @returns {Promise<{data: Object, publicUrl: string|null}>}
 */
async function uploadFile(file, destinationPath, options = {}) {
  const supabase = getSupabaseClient();
  const bucketName = options.bucket || 'uploads';
  const upsert = options.upsert !== undefined ? options.upsert : false;

  let fileBody = file;
  let remotePath = destinationPath;

  // Handle local file path string
  if (typeof file === 'string') {
    if (!fs.existsSync(file)) {
      throw new Error(`File not found at path: ${file}`);
    }
    fileBody = fs.readFileSync(file);
    if (!remotePath) {
      remotePath = path.basename(file);
    }
  }

  if (!remotePath) {
    throw new Error('destinationPath is required when uploading in-memory buffers or streams.');
  }

  const uploadOptions = {
    upsert,
    ...(options.contentType ? { contentType: options.contentType } : {})
  };

  try {
    const { data, error } = await supabase.storage
      .from(bucketName)
      .upload(remotePath, fileBody, uploadOptions);

    if (error) {
      throw error;
    }

    const { data: publicUrlData } = supabase.storage
      .from(bucketName)
      .getPublicUrl(remotePath);

    return {
      data,
      publicUrl: publicUrlData ? publicUrlData.publicUrl : null
    };
  } finally {
    // Explicitly release file buffer references so GC reclaims memory immediately
    fileBody = null;
  }
}

module.exports = {
  getSupabaseClient,
  uploadFile,
  upload: uploadFile
};

File: storage.py
import os
from typing import Optional, Union, BinaryIO
from supabase import create_client, Client


def get_supabase_client() -> Client:
    """
    Initializes and returns a Supabase client using environment variables.

    Requires:
        SUPABASE_URL: The URL of your Supabase project.
        SUPABASE_KEY: The API key (anon or service_role) for your Supabase project.

    Returns:
        Client: An instance of the Supabase Client.
    """
    supabase_url: Optional[str] = os.environ.get("SUPABASE_URL")
    supabase_key: Optional[str] = os.environ.get("SUPABASE_KEY")

    if not supabase_url or not supabase_key:
        raise ValueError(
            "SUPABASE_URL and SUPABASE_KEY environment variables must be set."
        )

    return create_client(supabase_url, supabase_key)


def upload_file(
    file: Union[str, bytes, BinaryIO],
    destination_path: Optional[str] = None,
    bucket_name: str = "uploads",
    content_type: Optional[str] = None,
    upsert: bool = False,
) -> dict:
    """
    Uploads a file to the specified Supabase storage bucket ('uploads' by default).

    Args:
        file: Either a file path (str), raw file bytes (bytes), or an open binary file object (BinaryIO).
        destination_path: The remote path/file name in the bucket. If not provided for a file path,
                          the base name of the local file is used.
        bucket_name: The Supabase storage bucket name (defaults to 'uploads').
        content_type: Optional MIME type of the file (e.g. 'application/pdf', 'image/png').
        upsert: Whether to overwrite the file if it already exists (defaults to False).

    Returns:
        dict: The response from the Supabase storage upload call.

    Raises:
        FileNotFoundError: If a file path string is given but the file does not exist.
        ValueError: If destination_path cannot be determined or credentials are missing.
    """
    supabase = get_supabase_client()

    file_options = {"upsert": "true" if upsert else "false"}
    if content_type:
        file_options["content-type"] = content_type

    # Case 1: Local file path
    if isinstance(file, str):
        if not os.path.isfile(file):
            raise FileNotFoundError(f"File not found at path: {file}")

        if destination_path is None:
            destination_path = os.path.basename(file)

        with open(file, "rb") as f:
            file_bytes = f.read()

        response = supabase.storage.from_(bucket_name).upload(
            path=destination_path,
            file=file_bytes,
            file_options=file_options,
        )
        return response

    # Case 2: In-memory bytes
    elif isinstance(file, bytes):
        if destination_path is None:
            raise ValueError(
                "destination_path must be specified when uploading raw bytes."
            )

        response = supabase.storage.from_(bucket_name).upload(
            path=destination_path,
            file=file,
            file_options=file_options,
        )
        return response

    # Case 3: BinaryIO / file-like object
    else:
        if destination_path is None:
            raise ValueError(
                "destination_path must be specified when uploading a file-like object."
            )

        file_bytes = file.read()
        response = supabase.storage.from_(bucket_name).upload(
            path=destination_path,
            file=file_bytes,
            file_options=file_options,
        )
        return response


if __name__ == "__main__":
    # Quick test / usage demonstration
    print("Supabase storage module loaded.")
